### ISSUE TITLE
threadAnalysis causing issues when forward declaring an async proc

### ISSUE THREAD

**Issue by mathsaey** [nim-lang/Nim#5699]
Combining `asyncdispatch` with threads seems to lead to some unintended behavior.
The following example will not compile unless  nim  is invoked with `--threadAnalysis:off`.

```nim
import asyncnet
import asyncdispatch

var
  peers {.threadVar.}: seq[AsyncSocket]
  thread: Thread[Port]

proc process(peer: AsyncSocket, remote: string) {.async.}

proc processHost(host: string) {.async.} = # template/generic instantiation from here
  let sock  = newAsyncSocket()
  await sock.connect(host, Port(12345))
  asyncCheck sock.process(host) # Error won't trigger when this call is removed

proc process(peer: AsyncSocket, remote: string) {.async.} =
  while true:
    let str = await peer.recvLine()
    if str == "":
      break
    else:
      asyncCheck processHost(str)

proc listenForPeers(port: Port) {.async.} =
  let listener = newAsyncSocket()
  listener.bindAddr(port)
  listener.listen()

  while true:
    let (host, peer) = await listener.acceptAddr()
    asyncCheck peer.process(host)
    peers.add(peer)

proc remoteThread(port: Port) {.thread.} =
  peers = @[]
  asyncCheck listenForPeers(port)
  runForever()

proc startRemoteHandler*(port: Port) =
  createThread(thread, remoteThread, port)
```

```
$ nim c --threads:on test.nim
...
test.nim(10, 32) template/generic instantiation from here
lib/pure/asyncmacro.nim(31, 8) Error: 'cb' is not GC-safe as it accesses 'nameIterVar' which is a global using GC'ed memory
```

```
$ nim c --threads:on --threadAnalysis:off test.nim
...
Hint: operation successful
```

**Comment by mathsaey** [comment]
Also discussed on IRC: https://irclogs.nim-lang.org/12-04-2017.html#17:40:55

**Comment by dom96** [comment]
@Araq Please take a look. I am stumped by this.

**Comment by mathsaey** [comment]
This seems to be solved in `17.0` when `gcSafe` is added to the forward declaration of `process`. Closing this.
